src/feature.js

```diff
@@ -1,147 +1,191 @@
 /* eslint no-unused-vars: ["error", { "varsIgnorePattern": "(feature)" }]*/
 
-/**  Example Feature module for a Shield Study.
- *
- *  UI:
- *  - during INSTALL only, show a notification bar with 2 buttons:
- *    - "Thanks".  Accepts the study (optional)
- *    - "I don't want this".  Uninstalls the study.
- *
- *  Firefox code:
- *  - Implements the 'introduction' to the 'button choice' study, via notification bar.
- *
- *  Demonstrates `studyUtils` API:
- *
- *  - `telemetry` to instrument "shown", "accept", and "leave-study" events.
- *  - `endStudy` to send a custom study ending.
- *
- **/
-class Feature {
-  constructor() {}
-  /** A Demonstration feature.
-   *
-   *  - variation: study info about particular client study variation
-   *  - reason: string of background.js install/startup/shutdown reason
-   *
-   */
-  configure(studyInfo) {
-    const feature = this;
-    const { variation, isFirstRun } = studyInfo;
+function generateRandomString() {
+  return Math.random().toString(36).substring(2, 15) +
+         Math.random().toString(36).substring(2, 15);
+}
+
+function isSupportedURLProtocol(url) {
+  return !!(url.match(/^(http(s?):\/\/)/im));
+}
+
+function getHostName(url) {
+  return url.match(/^(?:https?:\/\/)?(?:[^@\n]+@)?(?:www\.)?([^:\/\n]+)/im)[1];
```

Comment:
Could use the standard [URL API](https://developer.mozilla.org/en-US/docs/Web/API/URL) to tell if a URL is valid, extract the hostname etc. instead of having to do your own regex.

Reply:
In `about:preferences`, user can set any url into "Autoplay - Exception" which might include some formats which URL API doesn't accept. (eg. "123.com" is not a valid input format for URL API.)

Therefore, it means I need to do an additional transformation and check for every input, it would be more complicated. I would prefer to use my own regex rather than doing that.